### PR TITLE
improve the `defining a component` documentation

### PR DESCRIPTION
**Pull request by GustavoCaso**
### What does this PR do?

Reword some of the documentation around `Defining Components` doc.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: docs/components/defining-components.md [hunk 1]
```diff
@@ -170,9 +131,44 @@ func (m *mock) Foo(key string) string { ... }
 // AddedFoos implements Mock#AddedFoos.
 func (m *mock) AddedFoos() []Foo { ... }
 
-func newFoo(deps dependencies) Component {
+func newMockFoo(deps dependencies) Component {
```
Comment (hush-hush) [inline docs/components/defining-components.md:134]: We don't need to repeat the package name in its methods:
```suggestion
func newMock(deps dependencies) Component {
```

**Review by hush-hush — APPROVED** [review]

File: docs/components/defining-components.md [hunk 2]
```diff
@@ -1,30 +1,30 @@
 # Defining Components
 
+### You can use the invoke task `inv new-component comp/<bundleName>/<component>` to generate a scaffold for your new component.
```
Comment (urseberry) [inline docs/components/defining-components.md:3]:
```suggestion
You can use the invoke task `inv new-component comp/<bundleName>/<component>` to generate a scaffold for your new component.
```
This sentence instructs the user. It should not be a section heading.

File: docs/components/defining-components.md [hunk 3]
```diff
@@ -1,30 +1,30 @@
...
-      It should have a formulaic doc string like `// Component is the component type.`, deferring documentation to the package docs.
-      All interface methods should be exported and thoroughly documented.
-  * `impl/your_filename.go`      
-    * `Module` -- an `fx.Option` that can be included in the bundle's `Module` or an `fx.App` to make this component available. The `Module` is defined in a separate package from the component. This allows a package using only the component interface to not have to import the dependencies of the implementation of this component.
-      To assist with debugging, use `fxutil.Component(options...)`.
-      This item should have a formulaic doc string like `// Module defines the fx options for this component.`
+  * `comp/<bundleName>/<component>/<component>impl/<component>.go      
+    * `Module` -- an `fx.Option` that can be included in the bundle's `Module` or an `fx.App` to make this component available. 
+      The `Module` is defined in a separate package from the component, allowing a package to just import the interface without having to import that entire implementation..
```
Comment (urseberry) [inline docs/components/defining-components.md:20]:
```suggestion
      The `Module` is defined in a separate package from the component, allowing a package to import the interface without having to import the entire implementation.
```

File: docs/components/defining-components.md [hunk 4]
```diff
@@ -39,17 +39,12 @@ type Component interface {
 }
 ```
 
-The completed `impl/your_filename.go` looks like this:
-```go
-package config
+The Component interface is implemented in the file ``comp/<bundleName>/<component>/<component>impl/<component>.go`. Also, the `Module` definition.
```
Comment (urseberry) [inline docs/components/defining-components.md:42]:
```suggestion
The Component interface is implemented in the file `comp/<bundleName>/<component>/<component>impl/<component>.go` and the `Module` definition. 
```
The second sentence in this line is a fragment. I made my best guess as to the meaning and fixed it. Please make further edits if I missed the intended meaning.

File: docs/components/defining-components.md [hunk 5]
```diff
@@ -39,17 +39,12 @@ type Component interface {
...
-```go
-package config
+The Component interface is implemented in the file ``comp/<bundleName>/<component>/<component>impl/<component>.go`. Also, the `Module` definition. 
 
-// Module defines the fx options for this component.
-var Module = fxutil.Component(
-    fx.Provide(newFoo),
-)
-```
+The Module defintion function **must** be private.
```
Comment (urseberry) [inline docs/components/defining-components.md:44]:
```suggestion
The Module definition function **must** be private.
```

File: docs/components/defining-components.md [hunk 6]
```diff
@@ -39,17 +39,12 @@ type Component interface {
...
+
+The completed ``comp/<bundleName>/<component>/<component>impl/<component>.go` looks like this:
```
Comment (urseberry) [inline docs/components/defining-components.md:46]:
```suggestion
The completed `comp/<bundleName>/<component>/<component>impl/<component>.go` looks like this:
```

File: docs/components/defining-components.md [hunk 7]
```diff
@@ -170,9 +131,44 @@ func (m *mock) Foo(key string) string { ... }
...
+
+Detailed documentation of how to avoid bugs in using a component is an indicator of excessive complexity and should be treated as a bug.
+Simplifying the usage will improve the robustness of the Agent.
+
+Documentation should include:
+
+* Precise information on when each method may be called.
+  Can methods be called concurrently?
+  Are some methods invalid before the component has started?
+  Such assumptions are difficult to verify, so where possible try to make every method callable concurrently, at all times.
```
Comment (urseberry) [inline docs/components/defining-components.md:155]:
```suggestion
  Such assumptions are difficult to verify. Where possible, try to make every method callable concurrently, at all times.
```

**Review by urseberry — CHANGES_REQUESTED** [review]

File: docs/components/defining-components.md [hunk 8]
```diff
@@ -1,163 +1,110 @@
...
+    * `Component` -- The component interface.
+      This is the interface that other components can reference when declaring the component as a dependency via `fx`.
       It can be an empty interface, if there is no need for any methods.
-      It should have a formulaic doc string like `// Component is the component type.`, deferring documentation to the package docs.
-      All interface methods should be exported and thoroughly documented.
-  * `impl/your_filename.go`      
-    * `Module` -- an `fx.Option` that can be included in the bundle's `Module` or an `fx.App` to make this component available. The `Module` is defined in a separate package from the component. This allows a package using only the component interface to not have to import the dependencies of the implementation of this component.
-      To assist with debugging, use `fxutil.Component(options...)`.
-      This item should have a formulaic doc string like `// Module defines the fx options for this component.`
+  * `comp/<bundleName>/<component>/<component>impl/<component>.go
```
Comment (urseberry) [inline docs/components/defining-components.md:18]:
```suggestion
  * `comp/<bundleName>/<component>/<component>impl/<component>.go`
```

**Review by urseberry — APPROVED** [review]
Found one more formatting issue. Approving so I don't hold up the PR.

**Pull request merged**